apps/aechannel/src/aesc_window.erl

```diff
@@ -0,0 +1,117 @@
+-module(aesc_window).
+
+-export([new/0, new/1,
+         change_keep/2,
+         add/2,
+         pop/1,
+         keyfind/3,
+         keymember/3,
+         info_find/3,
+         to_list/1]).
+
+-export([record_fields/1]).
+
+-export_type([window/0]).
+
+-define(KEEP, 10).
+
+-type size()   :: non_neg_integer().
+-type entry()  :: any().
+
+-record(w, { n = 0         :: non_neg_integer()
+           , keep = ?KEEP  :: size()
+           , a = []        :: [entry()]
+           , b = []        :: [entry()]
+           }).
+
+-type window() :: #w{}.
+
+%% ==================================================================
+%% Tracing support
+record_fields(w) -> record_info(fields, w);
+record_fields(_) -> no.
+%% ==================================================================
+
+
+-spec new() -> window().
+new() ->
+    #w{}.
+
+-spec new(size()) -> window().
+new(Sz) when is_integer(Sz), Sz >= 0 ->
+    #w{keep = Sz}.
+
+-spec change_keep(size(), window()) -> window().
+change_keep(Keep, #w{} = W) when is_integer(Keep), Keep >= 0 ->
+    W#w{keep = Keep}.
+
+-spec add(entry(), window()) -> window().
+add(Item, #w{n = N, a = A, keep = Keep} = W) when N < Keep ->
+    W#w{n = N+1, a = [Item|A]};
+add(Item, #w{a = A} = W) ->
+    W#w{n = 1, a = [Item], b = A}.
+
+-spec pop(window()) -> {entry(), window()} | error.
+pop(#w{a = [], b = []}) ->
+    error;
+pop(#w{a = [], b = [H|T], n = N} = W) ->
+    {H, W#w{n = N-1, b = T}};
```

Comment:
This is popping one element out of `b`, and reducing `n`. So, based on this, it seems that `n` keeps `b` into account.

But in the second clause of `add/2`, the body `W#w{n = 1, a = [Item], b = A}` disregards the length of the new `b` in the computation of `n` (`n` seems to represent only `a` that is a list of length 1).

I still miss the full context/usage though seems suspicious.

A comment in the definition of field `n` in the record would ease understanding of its semantics.

Reply:
Correct, `n` should not be reduced when `pop()` picks from the `b` list, or the number of elements in `b` should be included in the total. Will consider a fix.

Reply:
I see addressed in abb8642